.cursor/rules/go-best-practices.mdc

```diff
@@ -0,0 +1,127 @@
+---
+description: 
+globs: *.go
+alwaysApply: false
+---
+# Go Best Practices
+
+Follow these best practices when writing Go code in this repository.
+
+## Core Principles
+
+- **Clarity and Simplicity**: Write code that is easy to read, understand, and maintain. Favor explicit over implicit.
+- **Dependency Injection**: Use dependency injection to decouple components and enable testing. The functional options pattern is the standard approach.
+- **Interface-Driven Design**: Define behavior through interfaces to enable mocking and loose coupling.
+- **Explicit Error Handling**: Handle all errors explicitly. Use structured error types when appropriate.
+
+## Architecture Patterns
+
+### Functional Options Pattern
+
+Use the functional options pattern for component initialization. This is the standard across controllers and main components.
+
+### Interface Design
+
+- **Small, Focused Interfaces**: Keep interfaces small and focused on specific behavior
+- **Interface Segregation**: Prefer multiple small interfaces over large ones
+- **Testability**: All external dependencies should be behind interfaces for mocking
+- **Naming**: Use descriptive names that indicate the behavior (e.g., `InstallationManager`, `NetUtils`)
+
+## Error Handling
+
+### Error Wrapping and Context
+
+- **Wrap Errors**: Always add context when propagating errors using `fmt.Errorf("operation failed: %w", err)`
+- **Use %w verb for error wrapping**: Use `%w` instead of `%v` when wrapping errors to maintain the error chain
+  ```go
+  return fmt.Errorf("failed to process: %w", err)  // Good
+  return fmt.Errorf("failed to process: %v", err)  // Bad
```

Comment:
i prefer it when we dont prefix the error with "failed to" or "unable to". this creates repetition and is overly verbose, eg `failed to install admin console: failed to helm install: failed to unmarshal helm values: the actual error`